bootstrap/shell/shell.go

```diff
@@ -84,6 +90,24 @@ func NewWithContext(ctx context.Context) (*Shell, error) {
 	return sh, nil
 }
 
+// WithStdin returns a copy of the Shell with the provided io.Reader set as the
+// Stdin for the next command. The copy should be discarded after one command.
+// e.g. sh.WithStdin(strings.NewReader("hello world")).Run("cat")
+func (s *Shell) WithStdin(r io.Reader) *Shell {
+	// cargo-culted cmdLock, not sure if it's needed
+	s.cmdLock.Lock()
+	defer s.cmdLock.Unlock()
+	// Can't copy struct like `newsh := *s` because sync.Mutex can't be copied.
+	return &Shell{
+		Logger: s.Logger,
+		Env:    s.Env,
+		stdin:  r, // our new stdin
+		Writer: s.Writer,
+		wd:     s.wd,
+		ctx:    s.ctx,
+	}
+}
```

Comment:
Here's the second-lowest layer learning to take Stdin — our Shell implementation.
The `stdin` eventually gets passed to the `process.Config`.